include/core/CStoredStringPtr.h

```diff
@@ -16,102 +16,14 @@
-//!
-//! IMPLEMENTATION DECISIONS:\n
-//! The private constructors make it hard to accidentally construct
-//! stored string pointers that are not managed by a string store.
-//!
-class CORE_EXPORT CStoredStringPtr {
-public:
-    //! NULL constructor.
-    CStoredStringPtr() noexcept;
-    ~CStoredStringPtr();
-
-    void swap(CStoredStringPtr& other) noexcept;
-
-    //! Get a reference to the string.
-    const std::string& operator*() const noexcept;
-
-    //! Get a pointer to the string.
-    const std::string* operator->() const noexcept;
-
-    //! Get a pointer to the string.
-    const std::string* get() const noexcept;
-
-    //! Is the pointer non-NULL?
-    explicit operator bool() const noexcept;
-
-    //! Is there only one pointer for this stored string?
-    //! This method is inefficient and should only be used in unit test
-    //! code (and then only infrequently).
-    bool isUnique() const noexcept;
-
-    //! Equality operator for NULL.
-    bool operator==(std::nullptr_t rhs) const noexcept;
-    bool operator!=(std::nullptr_t rhs) const noexcept;
-
-    //! Equality operator.
-    bool operator==(const CStoredStringPtr& rhs) const noexcept;
-    bool operator!=(const CStoredStringPtr& rhs) const noexcept;
-
-    //! Less than operator.
-    bool operator<(const CStoredStringPtr& rhs) const noexcept;
-
-    //! Claim memory usage is 0 in the main memory usage calculation, on the
-    //! assumption that the actual memory usage will be accounted for in a
-    //! string store.
-    static constexpr bool dynamicSizeAlwaysZero() { return true; }
-
-    //! Get the actual memory usage of the string.  For use by the string
-    //! store.
-    std::size_t actualMemoryUsage() const;
-    void debugActualMemoryUsage(const CMemoryUsage::TMemoryUsagePtr& mem) const;
-
-    //! These factory methods return a stored string pointer given a string.
-    //! They must only be used within string store classes that contain code
-    //! to account for memory usage outside of the main memory usage
-    //! calculation.
-    static CStoredStringPtr makeStoredString(const std::string& str);
-    static CStoredStringPtr makeStoredString(std::string&& str);
-
-private:
-    //! Non-NULL constructors are private to prevent accidental construction
-    //! outside of a string store.
-    explicit CStoredStringPtr(const std::string& str);
-    explicit CStoredStringPtr(std::string&& str);
-
-private:
-    using TStrCPtr = std::shared_ptr<const std::string>;
-
-    //! The wrapped shared_ptr.
-    TStrCPtr m_String;
-
-    friend CORE_EXPORT std::size_t hash_value(const CStoredStringPtr&);
-};
-
-//! Hash function named such that it will work automatically with Boost
-//! unordered containers.
-CORE_EXPORT
-std::size_t hash_value(const CStoredStringPtr& ptr);
-
-//! Swap for use by generic code.
-CORE_EXPORT
-void swap(CStoredStringPtr& lhs, CStoredStringPtr& rhs);
+// TODO(jan): rename
+using CStoredStringPtr = std::optional<std::string>;
```

Comment:
I did so. I think I've captured everything that can be null in an `optional<string>` and most things that cannot be null in `string` now. I tried to err on the safe side, so there probably are optionals in the code right now that are always set.

Reply:
@tveasey If you see `TOptionalStr`s that you think can be just `string`s, please let me know

Reply:
My inclination is you've done a reasonable effort first pass. I do suspect we may have some unnecessary nullable strings in a few places, but equally nothing you've left is trivial to unpick. I would be keen to get this through the overnights since it already simplifies a lot of code.